core/src/main/java/org/apache/spark/metrics/prometheus/client/exporter/PushGatewayWithTimestamp.java

```diff
@@ -0,0 +1,320 @@
+     */
+    @Deprecated
+    public void push(Collector collector, String job, String instance) throws IOException {
+        push(collector, job, Collections.singletonMap("instance", instance));
+    }
+
+    /**
+     * Pushes all metrics in a Collector,
+     * replacing only previously pushed metrics of the same name.
+     * <p>
+     * This is useful for pushing a single Gauge.
+     * <p>
+     * This uses the POST HTTP method.
+     * @deprecated use {@link #pushAdd(Collector, String, Map)}
+     */
+    @Deprecated
+    public void pushAdd(Collector collector, String job, String instance) throws IOException {
+        pushAdd(collector, job, Collections.singletonMap("instance", instance));
+    }
+
+    /**
+     * Deletes metrics from the Pushgateway.
+     * <p>
+     * This uses the DELETE HTTP method.
+     * @deprecated use {@link #delete(String, Map)}
+     */
+    @Deprecated
+    public void delete(String job, String instance) throws IOException {
+        delete(job, Collections.singletonMap("instance", instance));
+    }
+
+    void doRequest(CollectorRegistry registry, String job, Map<String,
+            String> groupingKey, String method, String timestamp) throws IOException {
+        String url = address + "/metrics/job/" + URLEncoder.encode(job, "UTF-8");
+        if (groupingKey != null) {
+            for (Map.Entry<String, String> entry: groupingKey.entrySet()) {
+                url += "/" + entry.getKey() + "/" + URLEncoder.encode(entry.getValue(), "UTF-8");
+            }
+        }
+
+        logger.info("Sending metrics data to '{}'", url);
+
+        HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
+        connection.setRequestProperty("Content-Type", TextFormatWithTimestamp.CONTENT_TYPE_004);
+        if (!method.equals("DELETE")) {
+            connection.setDoOutput(true);
+        }
+        connection.setRequestMethod(method);
+
+        connection.setConnectTimeout(10 * SECONDS_PER_MILLISECOND);
+        connection.setReadTimeout(10 * SECONDS_PER_MILLISECOND);
+        connection.connect();
+
+        try {
+            if (!method.equals("DELETE")) {
+                BufferedWriter writer =
+                        new BufferedWriter(
+                                new OutputStreamWriter(connection.getOutputStream(), "UTF-8"));
+                TextFormatWithTimestamp.write004(writer,
+                                                registry.metricFamilySamples(), timestamp);
+                writer.flush();
+                writer.close();
+            }
+
+            int response = connection.getResponseCode();
+            if (response != HttpURLConnection.HTTP_ACCEPTED) {
+                throw new IOException("Response code from " + url + " was " + response);
+            }
+        } catch (Exception ex) {
+            logger.error("Sending metrics failed due to: ", ex);
+        }
+
+        finally {
+            connection.disconnect();
```

Comment:
`connection` can be null if `new URL(url).openConnection()` at line 272 threw an exception.

Reply:
The `new URL(url).openConnection()` is outside of the `try-catch-block` this in case it throws an exception it exist `doRequest()` before reaching ```finally {
            connection.disconnect();```